File: src/test/java/fi/metatavu/metaform/test/functional/tests/AuditLogEntryTestsIT.java
package fi.metatavu.metaform.test.functional.tests;

import fi.metatavu.metaform.api.client.models.AuditLogEntry;
import fi.metatavu.metaform.api.client.models.Metaform;
import fi.metatavu.metaform.api.client.models.Reply;
import fi.metatavu.metaform.server.rest.ReplyMode;
import fi.metatavu.metaform.test.functional.AbstractIntegrationTest;
import fi.metatavu.metaform.test.functional.builder.TestBuilder;
import fi.metatavu.metaform.test.functional.builder.resources.KeycloakResource;
import fi.metatavu.metaform.test.functional.builder.resources.MysqlResource;
import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;


/**
 * Tests for AuditLogEntriesApi
 *
 * @author Katja Danilova
 */
@QuarkusTest
@QuarkusTestResource.List(value = {
  @QuarkusTestResource(MysqlResource.class),
  @QuarkusTestResource(KeycloakResource.class)
})
public class AuditLogEntryTestsIT extends AbstractIntegrationTest {

  @Test
  public void basicActionsOnReplyTest() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      Metaform parsedMetaform = builder.metaformAdmin().metaforms().readMetaform("simple");
      Metaform metaform = builder.metaformAdmin().metaforms().create(parsedMetaform);

      Map<String, Object> replyData = new HashMap<>();
      replyData.put("text", "Test text value");
      Reply createdReply = builder.test1().replies().createReplyWithData(replyData);

      Reply reply = builder.test1().replies().create(metaform.getId(), null, ReplyMode.REVISION.toString(), createdReply);
      builder.test1().replies().findReply(metaform.getId(), reply.getId(), null);
      builder.test1().replies().listReplies(metaform.getId(), null, null, null, null, null, true, null, null, null);
      builder.test1().replies().updateReply(metaform.getId(), reply.getId(), reply, reply.getOwnerKey());
      builder.test1().replies().delete(metaform.getId(), reply, reply.getOwnerKey());

      AuditLogEntry[] auditLogEntries = builder.test1().auditLogs().listAuditLogEntries(metaform.getId(), null, reply.getId(), null, null);

      Assertions.assertEquals(5, auditLogEntries.length);
      Assertions.assertEquals(String.format("user %s created reply %s", REALM1_USER_1_ID, reply.getId()), auditLogEntries[0].getMessage());
      Assertions.assertEquals(String.format("user %s viewed reply %s", REALM1_USER_1_ID, reply.getId()), auditLogEntries[1].getMessage());
      Assertions.assertEquals(String.format("user %s listed reply %s", REALM1_USER_1_ID, reply.getId()), auditLogEntries[2].getMessage());
      Assertions.assertEquals(String.format("user %s modified reply %s", REALM1_USER_1_ID, reply.getId()), auditLogEntries[3].getMessage());
      Assertions.assertEquals(String.format("user %s deleted reply %s", REALM1_USER_1_ID, reply.getId()), auditLogEntries[4].getMessage());
    }
  }

  @Test
  public void queryByUserTest() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      Metaform parsedMetaform = builder.metaformAdmin().metaforms().readMetaform("simple");
      Metaform metaform = builder.metaformAdmin().metaforms().create(parsedMetaform);

      Map<String, Object> replyData = new HashMap<>();
      replyData.put("text", "Test text value");
      Reply reply = builder.test1().replies().createReplyWithData(replyData);

      Reply createdReply1 = builder.test1().replies().create(metaform.getId(), null, ReplyMode.REVISION.toString(), reply);
      Reply createdReply2 = builder.test2().replies().create(metaform.getId(), null, ReplyMode.REVISION.toString(), reply);

      AuditLogEntry[] auditLogEntriesForUser1 = builder.test1().auditLogs().listAuditLogEntries(metaform.getId(), REALM1_USER_1_ID, null, null, null);
      AuditLogEntry[] auditLogEntriesForUser2 = builder.test1().auditLogs().listAuditLogEntries(metaform.getId(), REALM1_USER_2_ID, null, null, null);

      Assertions.assertEquals(String.format("user %s created reply %s", REALM1_USER_1_ID, createdReply1.getId()),
        auditLogEntriesForUser1[0].getMessage());
      Assertions.assertEquals(String.format("user %s created reply %s", REALM1_USER_2_ID, createdReply2.getId()),
        auditLogEntriesForUser2[0].getMessage());
    }
  }

  @Test
  public void queryByMetaformTest() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      Metaform parsedMetaform = builder.metaformAdmin().metaforms().readMetaform("simple");
      Metaform metaform1 = builder.metaformAdmin().metaforms().create(parsedMetaform);
      Metaform metaform2 = builder.metaformAdmin().metaforms().create(parsedMetaform);

      Map<String, Object> replyData = new HashMap<>();
      replyData.put("text", "Test text value");
      Reply reply = builder.test1().replies().createReplyWithData(replyData);

      Reply createdReply1 = builder.test1().replies().create(metaform1.getId(), null, ReplyMode.REVISION.toString(), reply);
      Reply createdReply2 = builder.test1().replies().create(metaform2.getId(), null, ReplyMode.REVISION.toString(), reply);

      AuditLogEntry[] metaform1AuditLogs = builder.test1().auditLogs().listAuditLogEntries(metaform1.getId(), null, null, null, null);
      AuditLogEntry[] metaform2AuditLogs = builder.test1().auditLogs().listAuditLogEntries(metaform2.getId(), null, null, null, null);

      Assertions.assertEquals(1, metaform1AuditLogs.length);
      Assertions.assertEquals(1, metaform2AuditLogs.length);
      Assertions.assertEquals(String.format("user %s created reply %s", REALM1_USER_1_ID, createdReply1.getId()), metaform1AuditLogs[0].getMessage());
      Assertions.assertEquals(String.format("user %s created reply %s", REALM1_USER_1_ID, createdReply2.getId()), metaform2AuditLogs[0].getMessage());
    }
  }

  /**
   * test verifies that sorting by reply id words
   *
   * @throws IOException
   */
  @Test
  public void queryByReplyIdTest() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      Metaform parsedMetaform = builder.metaformAdmin().metaforms().readMetaform("simple");
      Metaform metaform = builder.metaformAdmin().metaforms().create(parsedMetaform);

      Map<String, Object> replyData = new HashMap<>();
      replyData.put("text", "Test text value");
      Reply reply1 = builder.test1().replies().createReplyWithData(replyData);
      Reply reply2 = builder.test1().replies().createReplyWithData(replyData);

      Reply createdReply1 = builder.test1().replies().create(metaform.getId(), null, ReplyMode.REVISION.toString(), reply1);
      Reply createdReply2 = builder.test1().replies().create(metaform.getId(), null, ReplyMode.REVISION.toString(), reply2);

      AuditLogEntry[] auditLogEntries = builder.test1().auditLogs().listAuditLogEntries(metaform.getId(), null, null, null, null);
      Assertions.assertEquals(2, auditLogEntries.length);

      AuditLogEntry[] entryByReply = builder.test1().auditLogs().listAuditLogEntries(metaform.getId(), null, createdReply1.getId(), null, null);
      Assertions.assertEquals(1, entryByReply.length);
      Assertions.assertEquals(String.format("user %s created reply %s", REALM1_USER_1_ID, createdReply1.getId()), entryByReply[0].getMessage());
    }
  }

  @Test
  public void accessRightsTest() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      Metaform parsedMetaform = builder.metaformAdmin().metaforms().readMetaform("simple");
      Metaform metaform = builder.metaformAdmin().metaforms().create(parsedMetaform);

      Map<String, Object> replyData = new HashMap<>();
      replyData.put("text", "Test text value");
      Reply reply1 = builder.test1().replies().createReplyWithData(replyData);
      builder.test1().replies().create(metaform.getId(), null, ReplyMode.REVISION.toString(), reply1);

      AuditLogEntry[] auditLogEntries = builder.test1().auditLogs().listAuditLogEntries(metaform.getId(), null, null, null, null);
      Assertions.assertNotNull(auditLogEntries);

      builder.test2().auditLogs().assertListFailStatus(403, metaform.getId(), null, null, null, null);
    }
  }
}

File: src/test/java/fi/metatavu/metaform/server/test/functional/tests/AuditLogEntryTestsIT.kt
package fi.metatavu.metaform.server.test.functional.tests

import fi.metatavu.metaform.api.client.models.AuditLogEntry
import fi.metatavu.metaform.api.client.models.Metaform
import fi.metatavu.metaform.api.client.models.Reply
import fi.metatavu.metaform.server.rest.ReplyMode
import fi.metatavu.metaform.server.test.functional.AbstractTest
import fi.metatavu.metaform.server.test.functional.builder.PermissionScope
import fi.metatavu.metaform.server.test.functional.builder.TestBuilder
import fi.metatavu.metaform.server.test.functional.builder.auth.TestBuilderAuthentication
import fi.metatavu.metaform.server.test.functional.builder.resources.KeycloakResource
import fi.metatavu.metaform.server.test.functional.builder.resources.MysqlResource
import io.quarkus.test.common.QuarkusTestResource
import io.quarkus.test.junit.QuarkusTest
import io.quarkus.test.junit.TestProfile
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

/**
 * Tests for AuditLogEntriesApi
 *
 * @author Katja Danilova
 */
@QuarkusTest
@QuarkusTestResource.List(
        QuarkusTestResource(MysqlResource::class),
        QuarkusTestResource(KeycloakResource::class)
)
@TestProfile(GeneralTestProfile::class)
class AuditLogEntryTestsIT : AbstractTest() {
    @Test
    @Throws(Exception::class)
    fun basicActionsOnReply() {
        TestBuilder().use { builder ->
            val metaform: Metaform = builder.systemAdmin.metaforms.createFromJsonFile("simple")
            val replyData: MutableMap<String, Any> = HashMap()
            replyData["text"] = "Test text value"
            val createdReply: Reply = builder.test1.replies.createReplyWithData(replyData)
            val reply: Reply = builder.test1.replies.create(metaform.id!!, null, ReplyMode.REVISION.toString(), createdReply)
            builder.test1.replies.findReply(metaform.id, reply.id!!, null)
            builder.test1.replies.listReplies(metaform.id, null, null, null, null, null, true, null, null, null)
            builder.test1.replies.updateReply(metaform.id, reply.id, reply, reply.ownerKey)
            builder.test1.replies.delete(metaform.id, reply.id, reply.ownerKey)
            val auditLogEntries: Array<AuditLogEntry> = builder.test1.auditLogs.listAuditLogEntries(metaform.id, null, reply.id, null, null)
            Assertions.assertEquals(5, auditLogEntries.size)
            Assertions.assertEquals(java.lang.String.format("user %s created reply %s", REALM1_USER_1_ID, reply.id), auditLogEntries[0].message)
            Assertions.assertEquals(java.lang.String.format("user %s viewed reply %s", REALM1_USER_1_ID, reply.id), auditLogEntries[1].message)
            Assertions.assertEquals(java.lang.String.format("user %s listed reply %s", REALM1_USER_1_ID, reply.id), auditLogEntries[2].message)
            Assertions.assertEquals(java.lang.String.format("user %s modified reply %s", REALM1_USER_1_ID, reply.id), auditLogEntries[3].message)
            Assertions.assertEquals(java.lang.String.format("user %s deleted reply %s", REALM1_USER_1_ID, reply.id), auditLogEntries[4].message)
        }
    }

    @Test
    @Throws(Exception::class)
    fun queryByUser() {
        TestBuilder().use { builder ->
            val metaform: Metaform = builder.systemAdmin.metaforms.createFromJsonFile("simple")
            val replyData: MutableMap<String, Any> = HashMap()
            replyData["text"] = "Test text value"
            val replyWithData: Reply = builder.test1.replies.createReplyWithData(replyData)
            val reply1 = builder.test1.replies.create(metaform.id!!, null, ReplyMode.REVISION.toString(), replyWithData)
            val reply2 = builder.test2.replies.create(metaform.id, null, ReplyMode.REVISION.toString(), replyWithData)
            val auditLogEntriesForUser1: Array<AuditLogEntry> = builder.test1.auditLogs.listAuditLogEntries(metaform.id, REALM1_USER_1_ID, null, null, null)
            val auditLogEntriesForUser2: Array<AuditLogEntry> = builder.test1.auditLogs.listAuditLogEntries(metaform.id, REALM1_USER_2_ID, null, null, null)
            Assertions.assertEquals(java.lang.String.format("user %s created reply %s", REALM1_USER_1_ID, reply1.id),
                    auditLogEntriesForUser1[0].message)
            Assertions.assertEquals(java.lang.String.format("user %s created reply %s", REALM1_USER_2_ID, reply2.id),
                    auditLogEntriesForUser2[0].message)
        }
    }

    @Test
    @Throws(Exception::class)
    fun queryByMetaform() {
        TestBuilder().use { builder ->
            val metaform1: Metaform = builder.systemAdmin.metaforms.createFromJsonFile("simple")
            val metaform2: Metaform = builder.systemAdmin.metaforms.createFromJsonFile("simple")
            val replyData: MutableMap<String, Any> = HashMap()
            replyData["text"] = "Test text value"
            val replyWithData: Reply = builder.test1.replies.createReplyWithData(replyData)
            val reply1 = builder.test1.replies.create(metaform1.id!!, null, ReplyMode.REVISION.toString(), replyWithData)
            val reply2 = builder.test1.replies.create(metaform2.id!!, null, ReplyMode.REVISION.toString(), replyWithData)
            val metaform1AuditLogs: Array<AuditLogEntry> = builder.test1.auditLogs.listAuditLogEntries(metaform1.id, null, null, null, null)
            val metaform2AuditLogs: Array<AuditLogEntry> = builder.test1.auditLogs.listAuditLogEntries(metaform2.id, null, null, null, null)
            Assertions.assertEquals(1, metaform1AuditLogs.size)
            Assertions.assertEquals(1, metaform2AuditLogs.size)
            Assertions.assertEquals(java.lang.String.format("user %s created reply %s", REALM1_USER_1_ID, reply1.id), metaform1AuditLogs[0].message)
            Assertions.assertEquals(java.lang.String.format("user %s created reply %s", REALM1_USER_1_ID, reply2.id), metaform2AuditLogs[0].message)
        }
    }

    /**
     * test verifies that sorting by reply id words
     *
     * @throws IOException
     */
    @Test
    @Throws(Exception::class)
    fun queryByReplyId() {
        TestBuilder().use { builder ->
            val metaform: Metaform = builder.systemAdmin.metaforms.createFromJsonFile("simple")
            val replyData: MutableMap<String, Any> = HashMap()
            replyData["text"] = "Test text value"
            val replyWithData1: Reply = builder.test1.replies.createReplyWithData(replyData)
            val replyWithData2: Reply = builder.test1.replies.createReplyWithData(replyData)
            val reply3 = builder.test1.replies.create(metaform.id!!, null, ReplyMode.REVISION.toString(), replyWithData1)
            builder.test1.replies.create(metaform.id, null, ReplyMode.REVISION.toString(), replyWithData2)
            val auditLogEntries: Array<AuditLogEntry> = builder.test1.auditLogs.listAuditLogEntries(metaform.id, null, null, null, null)
            Assertions.assertEquals(2, auditLogEntries.size)
            val entryByReply: Array<AuditLogEntry> = builder.test1.auditLogs.listAuditLogEntries(metaform.id, null, reply3.id, null, null)
            Assertions.assertEquals(1, entryByReply.size)
            Assertions.assertEquals(java.lang.String.format("user %s created reply %s", REALM1_USER_1_ID, reply3.id), entryByReply[0].message)
        }
    }

    @Test
    @Throws(Exception::class)
    fun accessRights() {
        TestBuilder().use { builder ->
            val metaform: Metaform = builder.systemAdmin.metaforms.createFromJsonFile("simple")
            val replyData: MutableMap<String, Any> = HashMap()
            replyData["text"] = "Test text value"
            val reply1: Reply = builder.test1.replies.createReplyWithData(replyData)
            builder.test1.replies.create(metaform.id!!, null, ReplyMode.REVISION.toString(), reply1)
            val auditLogEntries: Array<AuditLogEntry> = builder.test1.auditLogs.listAuditLogEntries(metaform.id, null, null, null, null)
            Assertions.assertNotNull(auditLogEntries)
            builder.test2.auditLogs.assertListFailStatus(403, metaform.id, null, null, null, null)
        }
    }

    @Test
    @Throws(Exception::class)
    fun listAuditLogEntryPermission() {
        TestBuilder().use { testBuilder ->
            val metaform: Metaform = testBuilder.systemAdmin.metaforms.createFromJsonFile("simple")

            testBuilder.permissionTestByScopes(
                scope = PermissionScope.METAFORM_ADMIN,
                apiCaller = { authentication: TestBuilderAuthentication, _: Int ->
                    authentication.auditLogs.listAuditLogEntries(
                        metaformId = metaform.id!!,
                        userId = null,
                        replyId = null,
                        createdBefore = null,
                        createdAfter = null
                    )
                },
                metaformId = metaform.id
            )
        }
    }

    @Test
    @Throws(Exception::class)
    fun deleteAuditLogEntryPermission() {
        TestBuilder().use { testBuilder ->
            val metaform: Metaform = testBuilder.systemAdmin.metaforms.createFromJsonFile("simple")

            testBuilder.permissionTestByScopes(
                scope = PermissionScope.ANONYMOUS,
                apiCaller = { authentication: TestBuilderAuthentication, _: Int ->
                    val replyData: MutableMap<String, Any> = HashMap()
                    replyData["text"] = "Test text value"
                    val createdReply: Reply = testBuilder.test1.replies.createReplyWithData(replyData)
                    testBuilder.test1.replies.create(metaform.id!!, null, ReplyMode.REVISION.toString(), createdReply)

                    val auditLog = testBuilder.systemAdmin.auditLogs.listAuditLogEntries(
                        metaformId = metaform.id,
                        userId = null,
                        replyId = null,
                        createdBefore = null,
                        createdAfter = null
                    ).first()

                    authentication.auditLogs.deleteAuditLogsEntries(metaform.id, auditLog.id!!)
                }
            )
        }
    }
}